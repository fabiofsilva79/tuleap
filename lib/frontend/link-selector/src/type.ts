/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LinkSelectorItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface LinkSelector {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    destroy: () => void;
}

export type LinkSelectorSearchFieldCallback = (query: string) => void;

export type LinkSelectorSelectionCallback = (selected_value: unknown | null) => void;

export type LinkSelectorTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LinkSelectorItem
) => HTMLTemplateResult;

export interface LinkSelectorOptions {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly templating_callback: LinkSelectorTemplatingCallback;
    readonly selection_callback: LinkSelectorSelectionCallback;
    readonly search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
