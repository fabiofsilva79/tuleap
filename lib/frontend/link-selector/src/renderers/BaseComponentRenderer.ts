/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */
import type { LinkSelectorComponent } from "../type";
import { isLinkSelectorInAModal } from "../helpers/link-selector-in-modals-helper";

export class BaseComponentRenderer {
    constructor(
        private readonly doc: Document,
        private readonly source_select_box: HTMLSelectElement,
        private readonly placeholder: string,
        private readonly input_placeholder: string
    ) {}

    public renderBaseComponent(): LinkSelectorComponent {
        const wrapper_element = document.createElement("span");
        wrapper_element.classList.add("link-selector-component-wrapper");

        const link_selector_element = this.createLinkSelectorElement();
        const dropdown_element = this.createDropdownElement();
        const dropdown_list_element = this.createDropdownListElement();
        const selection_element = this.createSelectionElement();
        const placeholder_element = this.createPlaceholderElement();
        const search_field_element = this.createSearchFieldElement();

        selection_element.appendChild(placeholder_element);

        if (!this.source_select_box.disabled) {
            selection_element.setAttribute("tabindex", "0");
        }

        const search_section_element = this.createSearchSectionElement();
        search_section_element.appendChild(search_field_element);
        dropdown_element.appendChild(search_section_element);

        link_selector_element.appendChild(selection_element);
        dropdown_element.appendChild(dropdown_list_element);
        wrapper_element.appendChild(link_selector_element);
        this.doc.body.insertAdjacentElement("beforeend", dropdown_element);

        this.source_select_box.insertAdjacentElement("afterend", wrapper_element);

        if (isLinkSelectorInAModal(wrapper_element)) {
            dropdown_element.classList.add("link-selector-dropdown-over-modal");
        }

        return {
            wrapper_element,
            link_selector_element,
            dropdown_element,
            selection_element,
            placeholder_element,
            dropdown_list_element,
            search_field_element,
        };
    }

    private createDropdownListElement(): HTMLElement {
        const dropdown_list_element = document.createElement("ul");
        dropdown_list_element.classList.add("link-selector-dropdown-values-list");
        dropdown_list_element.setAttribute("role", "listbox");
        dropdown_list_element.setAttribute("aria-expanded", "false");
        dropdown_list_element.setAttribute("aria-hidden", "false");
        return dropdown_list_element;
    }

    private createDropdownElement(): HTMLElement {
        const dropdown_element = document.createElement("span");
        dropdown_element.classList.add("link-selector-dropdown");
        dropdown_element.setAttribute("data-test", "link-selector-dropdown");
        return dropdown_element;
    }

    private createPlaceholderElement(): Element {
        const placeholder_element = document.createElement("span");
        placeholder_element.classList.add("link-selector-placeholder");
        placeholder_element.appendChild(document.createTextNode(this.placeholder));
        return placeholder_element;
    }

    private createSelectionElement(): HTMLElement {
        const selection_element = document.createElement("span");
        selection_element.classList.add("link-selector-selection");
        selection_element.setAttribute("data-test", "link-selector-selection");
        selection_element.classList.add("link-selector-single");
        selection_element.setAttribute("role", "textbox");
        selection_element.setAttribute("aria-readonly", "true");

        return selection_element;
    }

    private createLinkSelectorElement(): Element {
        const list_picker_element = document.createElement("span");
        list_picker_element.classList.add("link-selector");

        if (this.source_select_box.disabled) {
            list_picker_element.classList.add("link-selector-disabled");
        }

        return list_picker_element;
    }

    private createSearchSectionElement(): Element {
        const search_section = document.createElement("span");
        search_section.classList.add("link-selector-single-dropdown-search-section");

        return search_section;
    }

    private createSearchFieldElement(): HTMLInputElement {
        const search_field_element = document.createElement("input");
        if (this.source_select_box.disabled) {
            search_field_element.setAttribute("disabled", "disabled");
        }
        search_field_element.setAttribute("data-test", "link-selector-search-field");
        search_field_element.classList.add("link-selector-search-field");
        search_field_element.setAttribute("type", "search");
        search_field_element.setAttribute("tabindex", "0");
        search_field_element.setAttribute("autocomplete", "off");
        search_field_element.setAttribute("autocorrect", "off");
        search_field_element.setAttribute("autocapitalize", "none");
        search_field_element.setAttribute("spellcheck", "none");
        search_field_element.setAttribute("role", "searchbox");
        search_field_element.setAttribute("aria-autocomplete", "list");
        search_field_element.setAttribute("aria-controls", "link-selector-dropdown-values-list");
        search_field_element.setAttribute("placeholder", this.input_placeholder);

        return search_field_element;
    }
}
