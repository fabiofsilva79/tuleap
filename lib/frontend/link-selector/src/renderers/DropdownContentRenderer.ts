/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { html, render } from "lit/html.js";
import type { HTMLTemplateResult } from "lit/html.js";
import type { ItemsMapManager } from "../items/ItemsMapManager";
import type { GroupCollection } from "../items/GroupCollection";
import { getGroupId } from "../helpers/group-id-helper";
import type { GroupOfItems } from "../items/GroupCollection";

export const getRenderedListItem = (
    option_id: string,
    template: HTMLTemplateResult,
    is_disabled: boolean
): Element => {
    let class_name = "link-selector-dropdown-option-value";
    if (is_disabled) {
        class_name = "link-selector-dropdown-option-value-disabled";
    }

    const document_fragment = document.createDocumentFragment();
    render(
        html`
            <li
                role="option"
                aria-selected="false"
                data-item-id="${option_id}"
                class="${class_name}"
                data-test="link-selector-item"
            >
                ${template}
            </li>
        `,
        document_fragment
    );

    const list_item = document_fragment.firstElementChild;
    if (list_item !== null) {
        return list_item;
    }

    throw new Error("Cannot render the list item");
};

const createEmptyDropdownState = (dropdown_message: string): HTMLTemplateResult =>
    html`
        <li
            class="link-selector-empty-dropdown-state"
            role="alert"
            aria-live="assertive"
            data-test="link-selector-empty-state"
        >
            ${dropdown_message}
        </li>
    `;

const getGroupLabel = (group: GroupOfItems): HTMLTemplateResult => {
    if (group.is_loading) {
        return html`
            <strong class="link-selector-group-label">
                ${group.label}
                <i
                    class="fa-solid fa-spin fa-circle-notch link-selector-loading-group-spinner"
                    data-test="link-selector-loading-group-spinner"
                ></i>
            </strong>
        `;
    }

    return html`<strong class="link-selector-group-label">${group.label}</strong>`;
};

const getGroupFooter = (group: GroupOfItems): HTMLTemplateResult =>
    group.footer_message !== "" && group.items.length > 0
        ? html`<p class="link-selector-group-footer" data-test="link-selector-group-footer">
              ${group.footer_message}
          </p>`
        : html``;

export class DropdownContentRenderer {
    constructor(
        private readonly dropdown_list_element: HTMLElement,
        private readonly items_map_manager: ItemsMapManager
    ) {}

    public renderLinkSelectorDropdownContent(groups: GroupCollection): void {
        const group_templates = groups.map((group) => {
            const group_id = getGroupId(group);
            const link_selector_items = this.items_map_manager
                .getLinkSelectorItems()
                .filter((item) => item.group_id === group_id);

            const items_template =
                group.items.length === 0
                    ? createEmptyDropdownState(group.empty_message)
                    : link_selector_items.map((item) => item.element);

            return html`
                <li class="link-selector-item-group">
                    ${getGroupLabel(group)}
                    <ul
                        role="group"
                        aria-label="${group.label}"
                        class="link-selector-dropdown-option-group"
                    >
                        ${items_template}
                    </ul>
                    ${getGroupFooter(group)}
                </li>
            `;
        });

        render(group_templates, this.dropdown_list_element);
    }
}
