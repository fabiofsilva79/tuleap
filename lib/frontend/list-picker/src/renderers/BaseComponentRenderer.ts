/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */
import type { ListPickerComponent, ListPickerOptions } from "../type";
import { isListPickerInAModal } from "../helpers/list-picker-in-modals-helper";

export class BaseComponentRenderer {
    constructor(
        private readonly doc: HTMLDocument,
        private readonly source_select_box: HTMLSelectElement,
        private readonly options?: ListPickerOptions
    ) {}

    public renderBaseComponent(): ListPickerComponent {
        const wrapper_element = document.createElement("span");
        wrapper_element.classList.add("list-picker-component-wrapper");
        wrapper_element.setAttribute("data-list-picker", "wrapper");

        const list_picker_element = this.createListPickerElement();
        const dropdown_element = this.createDropdownElement();
        const dropdown_list_element = this.createDropdownListElement();
        const selection_element = this.createSelectionElement();
        const placeholder_element = this.createPlaceholderElement();
        const search_field_element = this.createSearchFieldElement();

        let search_section_multiple: Element | null = null;
        if (this.source_select_box.multiple) {
            search_field_element.setAttribute("placeholder", this.options?.placeholder ?? "");
            const new_search_section = this.createSearchSectionForMultipleListPicker();
            new_search_section.appendChild(search_field_element);
            selection_element.appendChild(new_search_section);
            search_section_multiple = new_search_section;
        } else {
            selection_element.appendChild(placeholder_element);

            if (this.options?.is_filterable) {
                const search_section_element = this.createSearchSectionElement();
                search_section_element.appendChild(search_field_element);
                dropdown_element.appendChild(search_section_element);
            }
        }

        list_picker_element.appendChild(selection_element);
        dropdown_element.appendChild(dropdown_list_element);
        wrapper_element.appendChild(list_picker_element);

        const element_attributes_updater = this.buildListPickerAttributesUpdater(
            list_picker_element,
            search_field_element,
            selection_element,
            search_section_multiple
        );
        element_attributes_updater();

        this.doc.body.insertAdjacentElement("beforeend", dropdown_element);

        this.source_select_box.insertAdjacentElement("afterend", wrapper_element);

        if (isListPickerInAModal(wrapper_element)) {
            dropdown_element.classList.add("list-picker-dropdown-over-modal");
        }

        return {
            wrapper_element,
            list_picker_element,
            dropdown_element,
            selection_element,
            placeholder_element,
            dropdown_list_element,
            search_field_element,
            element_attributes_updater,
        };
    }

    private createSearchSectionForMultipleListPicker(): Element {
        const search_section = document.createElement("span");

        search_section.classList.add("list-picker-multiple-search-section");
        return search_section;
    }

    private createDropdownListElement(): Element {
        const dropdown_list_element = document.createElement("ul");
        dropdown_list_element.classList.add("list-picker-dropdown-values-list");
        dropdown_list_element.setAttribute("role", "listbox");
        dropdown_list_element.setAttribute("aria-expanded", "false");
        dropdown_list_element.setAttribute("aria-hidden", "false");
        return dropdown_list_element;
    }

    private createDropdownElement(): HTMLElement {
        const dropdown_element = document.createElement("span");
        dropdown_element.classList.add("list-picker-dropdown");
        return dropdown_element;
    }

    private createPlaceholderElement(): Element {
        const placeholder_element = document.createElement("span");
        placeholder_element.classList.add("list-picker-placeholder");
        placeholder_element.appendChild(document.createTextNode(this.options?.placeholder ?? ""));
        return placeholder_element;
    }

    private createSelectionElement(): HTMLElement {
        const selection_element = document.createElement("span");
        selection_element.classList.add("list-picker-selection");
        selection_element.setAttribute("data-test", "list-picker-selection");

        if (this.source_select_box.multiple) {
            selection_element.classList.add("list-picker-multiple");
            selection_element.setAttribute("aria-haspopup", "true");
            selection_element.setAttribute("aria-expanded", "false");
            selection_element.setAttribute("role", "combobox");
        } else {
            selection_element.classList.add("list-picker-single");
            selection_element.setAttribute("role", "textbox");
            selection_element.setAttribute("aria-readonly", "true");
        }
        return selection_element;
    }

    private createListPickerElement(): Element {
        const list_picker_element = document.createElement("span");
        list_picker_element.classList.add("list-picker");

        if (this.source_select_box.multiple) {
            list_picker_element.classList.add("list-picker-in-multiple-mode");
        }

        return list_picker_element;
    }

    private buildListPickerAttributesUpdater(
        list_picker_element: Element,
        search_field_element: HTMLInputElement,
        selection_element: HTMLElement,
        search_section_multiple: Element | null
    ): () => void {
        return (): void => {
            const LIST_PICKER_DISABLED_CLASS = "list-picker-disabled";
            const LIST_PICKER_MULTIPLE_SEARCH_SECTION_DISABLED_CLASS =
                "list-picker-multiple-search-section-disabled";

            if (this.source_select_box.disabled) {
                list_picker_element.classList.add(LIST_PICKER_DISABLED_CLASS);
                search_field_element.setAttribute("disabled", "disabled");
                search_section_multiple?.classList.add(
                    LIST_PICKER_MULTIPLE_SEARCH_SECTION_DISABLED_CLASS
                );
            } else {
                list_picker_element.classList.remove(LIST_PICKER_DISABLED_CLASS);
                search_field_element.removeAttribute("disabled");
                search_section_multiple?.classList.remove(
                    LIST_PICKER_MULTIPLE_SEARCH_SECTION_DISABLED_CLASS
                );
            }
            this.updateSelectionElementAttributes(selection_element);
        };
    }

    private updateSelectionElementAttributes(selection_element: HTMLElement): void {
        if (this.source_select_box.multiple) {
            if (this.source_select_box.disabled) {
                selection_element.setAttribute("aria-disabled", "true");
            } else {
                selection_element.setAttribute("aria-disabled", "false");
            }
        }

        if (this.source_select_box.disabled) {
            selection_element.removeAttribute("tabindex");
        } else {
            if (this.source_select_box.multiple) {
                selection_element.setAttribute("tabindex", "-1");
            } else {
                selection_element.setAttribute("tabindex", "0");
            }
        }
    }

    private createSearchSectionElement(): Element {
        const search_section = document.createElement("span");
        search_section.classList.add("list-picker-single-dropdown-search-section");

        return search_section;
    }

    private createSearchFieldElement(): HTMLInputElement {
        const search_field_element = document.createElement("input");
        search_field_element.setAttribute("data-test", "list-picker-search-field");
        search_field_element.classList.add("list-picker-search-field");
        search_field_element.setAttribute("type", "search");
        search_field_element.setAttribute("tabindex", "0");
        search_field_element.setAttribute("autocomplete", "off");
        search_field_element.setAttribute("autocorrect", "off");
        search_field_element.setAttribute("autocapitalize", "none");
        search_field_element.setAttribute("spellcheck", "none");
        search_field_element.setAttribute("role", "searchbox");
        search_field_element.setAttribute("aria-autocomplete", "list");
        search_field_element.setAttribute("aria-controls", "list-picker-dropdown-values-list");

        return search_field_element;
    }
}
