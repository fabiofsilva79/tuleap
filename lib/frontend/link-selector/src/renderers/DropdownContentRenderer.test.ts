/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { describe, it, expect, beforeEach } from "vitest";
import { html } from "lit/html.js";
import { DropdownContentRenderer, getRenderedListItem } from "./DropdownContentRenderer";
import { BaseComponentRenderer } from "./BaseComponentRenderer";
import { ItemsMapManager } from "../items/ItemsMapManager";
import { ListItemMapBuilder } from "../items/ListItemMapBuilder";
import { GroupCollectionBuilder } from "../../tests/builders/GroupCollectionBuilder";
import type { GroupCollection } from "../items/GroupCollection";
import { TemplatingCallbackStub } from "../../tests/stubs/TemplatingCallbackStub";

describe("DropDownContentRenderer", () => {
    let dropdown: Element, dropdown_list: HTMLElement;

    const render = (groups: GroupCollection): void => {
        const items_map_manager = new ItemsMapManager(
            ListItemMapBuilder(TemplatingCallbackStub.build())
        );
        items_map_manager.refreshItemsMap(groups);
        const renderer = new DropdownContentRenderer(dropdown_list, items_map_manager);
        return renderer.renderLinkSelectorDropdownContent(groups);
    };

    describe("rendering", () => {
        beforeEach(() => {
            const doc = document.implementation.createHTMLDocument();
            const select = doc.createElement("select");
            const { dropdown_element, dropdown_list_element } = new BaseComponentRenderer(
                document.implementation.createHTMLDocument(),
                select,
                "",
                ""
            ).renderBaseComponent();

            dropdown = dropdown_element;
            dropdown_list = dropdown_list_element;
        });

        it("renders grouped list items", () => {
            render(GroupCollectionBuilder.withTwoGroups());
            expect(stripExpressionComments(dropdown.innerHTML)).toMatchSnapshot();
        });

        it("renders empty option groups when they have a placeholder", () => {
            const empty_state_text = "No results found on the server";
            const groups = GroupCollectionBuilder.withSingleGroup({
                empty_message: empty_state_text,
                items: [],
            });

            render(groups);

            const empty_state = dropdown_list.querySelector(
                "[data-test=link-selector-empty-state]"
            );
            if (!empty_state) {
                throw new Error("The empty state has not been found in the dropdown");
            }
            expect(empty_state.textContent?.trim()).toBe(empty_state_text);
        });

        it(`renders group footer message below the list of items`, () => {
            const footer_message = "Maybe there are more results";
            const groups = GroupCollectionBuilder.withSingleGroup({
                items: [
                    { value: { id: 1 }, is_disabled: false },
                    { value: { id: 2 }, is_disabled: false },
                ],
                footer_message,
            });

            render(groups);

            const footer = dropdown_list.querySelector("[data-test=link-selector-group-footer]");
            expect(footer?.textContent?.trim()).toBe(footer_message);
        });

        it("renders a spinner next to the group title when it is loading", () => {
            const empty_state_text = "I am loading, wait a second!";
            const groups = GroupCollectionBuilder.withSingleGroup({
                label: "A group still loading",
                empty_message: empty_state_text,
                items: [],
                is_loading: true,
            });

            render(groups);

            const spinner = dropdown_list.querySelector(
                "[data-test=link-selector-loading-group-spinner]"
            );
            expect(spinner).toBeDefined();
        });
    });

    describe(`getRenderedListItem()`, () => {
        it(`renders a list item from a given template`, () => {
            const item = getRenderedListItem(
                "value_1",
                html`
                    <span>Badge</span>
                    Value 1
                `,
                false
            );

            expect(stripExpressionComments(item.outerHTML)).toMatchInlineSnapshot(`
                <li role="option" aria-selected="false" data-test="link-selector-item" data-item-id="value_1" class="link-selector-dropdown-option-value">

                  <span>Badge</span>
                  Value 1

                </li>
            `);
        });

        it(`renders a disabled list item`, () => {
            const item = getRenderedListItem(
                "value_1",
                html`
                    <span>Badge</span>
                    Value 1
                `,
                true
            );

            expect(item.classList.contains("link-selector-dropdown-option-value-disabled")).toBe(
                true
            );
        });
    });
});

/**
 * See https://github.com/lit/lit/blob/lit%402.0.2/packages/lit-html/src/test/test-utils/strip-markers.ts
 */
function stripExpressionComments(html: string): string {
    return html.replace(/<!--\?lit\$[0-9]+\$-->|<!--\??-->/g, "");
}
