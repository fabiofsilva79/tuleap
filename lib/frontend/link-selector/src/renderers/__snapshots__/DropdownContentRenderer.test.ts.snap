// Vitest Snapshot v1

exports[`DropDownContentRenderer > rendering > renders grouped list items 1`] = `
<span class="link-selector-single-dropdown-search-section"><input data-test="link-selector-search-field" class="link-selector-search-field" type="search" tabindex="0" autocomplete="off" autocorrect="off" autocapitalize="none" spellcheck="none" role="searchbox" aria-autocomplete="list" aria-controls="link-selector-dropdown-values-list" placeholder=""></span>
<ul class="link-selector-dropdown-values-list" role="listbox" aria-expanded="false" aria-hidden="false">
  <li class="link-selector-item-group">
    <strong class="link-selector-group-label">Group 1</strong>
    <ul role="group" class="link-selector-dropdown-option-group" aria-label="Group 1">
      <li role="option" aria-selected="false" data-test="link-selector-item" data-item-id="link-selector-item-group1-0" class="link-selector-dropdown-option-value">
        Value 0
      </li>
      <li role="option" aria-selected="false" data-test="link-selector-item" data-item-id="link-selector-item-group1-1" class="link-selector-dropdown-option-value">
        Value 1
      </li>
      <li role="option" aria-selected="false" data-test="link-selector-item" data-item-id="link-selector-item-group1-2" class="link-selector-dropdown-option-value">
        Value 2
      </li>
    </ul>

  </li>

  <li class="link-selector-item-group">
    <strong class="link-selector-group-label">Group 2</strong>
    <ul role="group" class="link-selector-dropdown-option-group" aria-label="Group 2">
      <li role="option" aria-selected="false" data-test="link-selector-item" data-item-id="link-selector-item-group2-3" class="link-selector-dropdown-option-value">
        Value 3
      </li>
      <li role="option" aria-selected="false" data-test="link-selector-item" data-item-id="link-selector-item-group2-4" class="link-selector-dropdown-option-value">
        Value 4
      </li>
      <li role="option" aria-selected="false" data-test="link-selector-item" data-item-id="link-selector-item-group2-5" class="link-selector-dropdown-option-value-disabled">
        Value 5
      </li>
    </ul>

  </li>
</ul>
`;
